photon-lib/py/photonlibpy/packet.py

```diff
@@ -198,3 +198,117 @@ def decodeOptional(self, serde: Type) -> Optional[Any]:
+        """
+        self.packetData = self.packetData + struct.pack(packFormat, value)
+        self.size = len(self.packetData)
+
+    def encode8(self, value: int):
+        """
+        Encodes a single byte and appends it to the packet.
+        """
+        self._encodeGeneric("<b", value)
+
+    def encode16(self, value: int):
+        """
+        Encodes a short (2 bytes) and appends it to the packet.
+        """
+        self._encodeGeneric("<h", value)
+
+    def encodeInt(self, value: int):
+        """
+        Encodes an int (4 bytes) and appends it to the packet.
+        """
+        self._encodeGeneric("<l", value)
+
+    def encodeFloat(self, value: float):
+        """
+        Encodes a float (4 bytes) and appends it to the packet.
+        """
+        self._encodeGeneric("<f", value)
+
+    def encodeLong(self, value: int):
+        """
+        Encodes a long (8 bytes) and appends it to the packet.
+        """
+        self._encodeGeneric("<q", value)
+
+    def encodeDouble(self, value: float):
+        """
+        Encodes a double (8 bytes) and appends it to the packet.
+        """
+        self._encodeGeneric("<d", value)
+
+    def encodeBoolean(self, value: bool):
+        """
+        Encodes a boolean as a single byte and appends it to the packet.
+        """
+        self.encode8(1 if value else 0)
+
+    def encodeDoubleArray(self, values: list[float]):
+        """
+        Encodes an array of doubles and appends it to the packet.
+        """
+        self.encode8(len(values))
+        for value in values:
+            self.encodeDouble(value)
+
+    def encodeShortList(self, values: list[int]):
+        """
+        Encodes a list of shorts, with length prefixed as a single byte.
+        """
+        self.encode8(len(values))
+        for value in values:
+            self.encode16(value)
+
+    def encodeTransform(self, transform: Transform3d):
+        """
+        Encodes a Transform3d (translation and rotation) and appends it to the packet.
+        """
+        # Encode Translation3d part (x, y, z)
+        self.encodeDouble(transform.translation().x)
+        self.encodeDouble(transform.translation().y)
+        self.encodeDouble(transform.translation().z)
+
+        # Encode Rotation3d as Quaternion (w, x, y, z)
+        quaternion = transform.rotation().getQuaternion()
+        self.encodeDouble(quaternion.W())
+        self.encodeDouble(quaternion.X())
+        self.encodeDouble(quaternion.Y())
+        self.encodeDouble(quaternion.Z())
+
+    def encodeList(self, values: list[Any], serde: Type):
+        """
+        Encodes a list of items using a specific serializer and appends it to the packet.
+        """
+        # There is a bad interaction going on with PhotonTrackedTargetSerde.minAreaRectCorners.
+        # It can be passed as a None type but this isnt supposed to be handled here with nt
+        # It should just have 0 size from default initialisation instead....
```

Comment:
Our type hints here claim that the list is never null, just empty at worst. Should we track down the None, or just make the type optional?

Reply:
I am not sure. This code is correct at the moment IMO.

The problem is `PhotonTrackedTarget` over [here](https://github.com/PhotonVision/photonvision/blob/master/photon-lib/py/photonlibpy/targeting/photonTrackedTarget.py).  The list types should not be optional from what I understand. I think it is totally valid for them to just be empty lists to convey the same things as what None is doing.

Part of me wants the sim stuff in as quickly as possible. I know my team has been wanting it for a long time and this PR is blocking that at the moment. The other part of me wants to fix the root cause of the problem because its the better solution :p.

Reply:
So looking at the C++ Code:

``` C++
struct PhotonTrackedTarget_PhotonStruct {
  double yaw;
  double pitch;
  double area;
  double skew;
  int32_t fiducialId;
  int32_t objDetectId;
  float objDetectConf;
  frc::Transform3d bestCameraToTarget;
  frc::Transform3d altCameraToTarget;
  double poseAmbiguity;
  std::vector<photon::TargetCorner> minAreaRectCorners;
  std::vector<photon::TargetCorner> detectedCorners;

  friend bool operator==(PhotonTrackedTarget_PhotonStruct const&,
                         PhotonTrackedTarget_PhotonStruct const&) = default;
};
```

Those vectors are not optional so these lists shouldn't be either. That is the correct solution but IDK if we are ok with breaking that if people depend on it

Reply:
Yep agreed with changing to empty lists now. The time to make these changes is before 2025 and this change is super tiny. seeeeeeeeeeeeeend it

Reply:
See #1539 for the conclusion to this discussion